yolo_to_yolo/data_transformers.py

```diff
@@ -47,3 +47,137 @@ def __call__(self, input_data: YoloImageData) -> Iterable[YoloImageData]:
+                y1 = max(y1 - self.min_padding, 0)
+                x2 = min(x1 + width + self.min_padding * 2, img_width)
+                y2 = min(y1 + height + self.min_padding * 2, img_height)
+
+                # Apply crop
+                cropped_image = input_data.image[y1:y2, x1:x2]
+
+                # print(f"Cropped Image Shape for {shape_label.classname}: {cropped_image.shape}")
+
+                # Create new labels for the cropped image
+                new_labels = [YoloLabel(location=YoloBbox(x=0.5, y=0.5, w=1.0, h=1.0), classname=shape_label.classname)]
+                if best_char_label:
+                    new_labels.append(YoloLabel(location=YoloBbox(x=0.5, y=0.5, w=1.0, h=1.0), classname=best_char_label.classname))
+
+                # Make new image data to yield
+                new_img_data = YoloImageData(
+                    img_id=f"{input_data.img_id}_{shape_label.classname}",
+                    task=input_data.task,
+                    image=cropped_image,
+                    labels=new_labels
+                )
+
+                yield new_img_data
+
+    def _calculate_iou(self, bbox1: YoloBbox, bbox2: YoloBbox) -> float:
+        """
+        Calculate Intersection over Union (IoU) for two bounding boxes (this is basically an "overlap score")
+        """
+        x1_min, y1_min = bbox1.x - bbox1.w / 2, bbox1.y - bbox1.h / 2
+        x1_max, y1_max = bbox1.x + bbox1.w / 2, bbox1.y + bbox1.h / 2
+        x2_min, y2_min = bbox2.x - bbox2.w / 2, bbox2.y - bbox2.h / 2
+        x2_max, y2_max = bbox2.x + bbox2.w / 2, bbox2.y + bbox2.h / 2
+
+        intersect_x_min = max(x1_min, x2_min)
+        intersect_y_min = max(y1_min, y2_min)
+        intersect_x_max = min(x1_max, x2_max)
+        intersect_y_max = min(y1_max, y2_max)
+
+        if intersect_x_min < intersect_x_max and intersect_y_min < intersect_y_max:
+            intersection_area = (intersect_x_max - intersect_x_min) * (intersect_y_max - intersect_y_min)
+            bbox1_area = (x1_max - x1_min) * (y1_max - y1_min)
+            bbox2_area = (x2_max - x2_min) * (y2_max - y2_min)
+            iou = intersection_area / float(bbox1_area + bbox2_area - intersection_area)
+            return iou
+        return 0.0
+
+    def _get_combined_bbox(self, shape_bbox: YoloBbox, char_label: YoloLabel, img_width: int, img_height: int) -> tuple[int, int, int, int]:
+        """
+        Get the smallest bbox that contains both the shape and character bboxes
+        """
+        x_center, y_center, bbox_width, bbox_height = shape_bbox
+        x1 = int((x_center - bbox_width / 2) * img_width)
+        y1 = int((y_center - bbox_height / 2) * img_height)
+        x2 = int((x_center + bbox_width / 2) * img_width)
+        y2 = int((y_center + bbox_height / 2) * img_height)
+
+        if char_label:
+            char_bbox = char_label.location
+            char_x_center, char_y_center, char_bbox_width, char_bbox_height = char_bbox
```

Comment:
This is throwing a type error because the `location` can be a bbox or outline. If you don't want to implement this transformer for the outline yet, you can just check `isinstance` and throw a `NotImplementedError`.

Reply:
Not sure if you have Python type-checking turned on in the IDE, but you should.